support for Illumos/Solaris

## Output of `restic version`

I was trying to build latest stable release, 0.8.1.

## How did you run restic exactly?

I didn't run it, I was only trying to build it.

## What backend/server/service did you use to store the repository?

macOS, build was done with reproducible builder via docker image.

## Expected behavior

I would love to see restic building properly on Illumos/Solaris OS.

## Actual behavior

It fails with the following error messages:

```
docker run --rm --volume $PWD:/home/build restic/builder v0.8.1.tar.gz
Go version is go version go1.9.2 linux/amd64
path is /tmp/restic-release-JzlHL4
version is 0.8.1
outputdir is /home/build/restic-0.8.1-20180124-104215
restic_0.8.1_solaris_amd64
# github.com/restic/restic/vendor/bazil.org/fuse
github.com/restic/restic/vendor/bazil.org/fuse/error_std.go:27:20: undefined: errNoXattr
github.com/restic/restic/vendor/bazil.org/fuse/fuse.go:1345:26: undefined: attr
github.com/restic/restic/vendor/bazil.org/fuse/fuse_kernel.go:166:28: undefined: syscall.O_DIRECTORY
github.com/restic/restic/vendor/bazil.org/fuse/fuse_kernel.go:404:17: undefined: attr
# github.com/restic/restic/internal/restic
github.com/restic/restic/internal/restic/node.go:231:14: node.restoreSymlinkTimestamps undefined (type Node has no field or method restoreSymlinkTimestamps)
github.com/restic/restic/internal/restic/node_unix.go:21:18: cannot use statUnix(*s) (type statUnix) as type statT in return argument:
        statUnix does not implement statT (missing atim method)
github.com/restic/restic/internal/restic/node_xattr.go:16:10: undefined: xattr.Get
github.com/restic/restic/internal/restic/node_xattr.go:26:10: undefined: xattr.List
github.com/restic/restic/internal/restic/node_xattr.go:35:7: undefined: xattr.Set
build failed: exit status 2
exit status 1
```

## Steps to reproduce the behavior

Try to build restic for solaris/amd64 platform via https://github.com/restic/builder

## Do you have any idea what may have caused this?

Likely lack of support for Solaris platform.

## Do you have an idea how to solve the issue?

No, but I would hope having OpenBSD and FreeBSD support in place already, that it won't prove super difficult to include one more. Especially knowing that rest-server does support it.

## Did restic help you or made you happy in any way?

I honestly think it's the best backup solution out there and I would love to have it running on my SmartOS machines.

--- a/changelog/0.8.3_2018-02-26/pull-1649
+++ b/changelog/0.8.3_2018-02-26/pull-1649
@@ -1,0 +1,3 @@
+Enhancement: Add illumos/Solaris support.
+
+https://github.com/restic/restic/pull/1649

--- a/cmd/restic/cmd_mount.go
+++ b/cmd/restic/cmd_mount.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package main

--- a/cmd/restic/integration_fuse_test.go
+++ b/cmd/restic/integration_fuse_test.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package main

--- a/doc/050_restore.rst
+++ b/doc/050_restore.rst
@@ -63,7 +63,8 @@ command to serve the repository with FUSE:
     Now serving /tmp/backup at /mnt/restic
     Don't forget to umount after quitting!
 
-Mounting repositories via FUSE is not possible on Windows and OpenBSD.
+Mounting repositories via FUSE is not possible on OpenBSD, Solaris/illumos
+and Windows.
 
 Restic supports storage and preservation of hard links. However, since
 hard links exist in the scope of a filesystem by definition, restoring

--- a/internal/backend/sftp/foreground_solaris.go
+++ b/internal/backend/sftp/foreground_solaris.go
@@ -1,0 +1,25 @@
+package sftp
+
+import (
+	"os/exec"
+	"syscall"
+
+	"github.com/restic/restic/internal/errors"
+)
+
+func startForeground(cmd *exec.Cmd) (bg func() error, err error) {
+	// run the command in it's own process group so that SIGINT
+	// is not sent to it.
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid: true,
+	}
+
+	// start the process
+	err = cmd.Start()
+	if err != nil {
+		return nil, errors.Wrap(err, "cmd.Start")
+	}
+
+	bg = func() error { return nil }
+	return bg, nil
+}

--- a/internal/backend/sftp/foreground_unix.go
+++ b/internal/backend/sftp/foreground_unix.go
@@ -1,3 +1,4 @@
+// +build !solaris
 // +build !windows
 
 package sftp

--- a/internal/fuse/blob_size_cache.go
+++ b/internal/fuse/blob_size_cache.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package fuse

--- a/internal/fuse/dir.go
+++ b/internal/fuse/dir.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package fuse

--- a/internal/fuse/file.go
+++ b/internal/fuse/file.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package fuse

--- a/internal/fuse/file_test.go
+++ b/internal/fuse/file_test.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package fuse

--- a/internal/fuse/link.go
+++ b/internal/fuse/link.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package fuse

--- a/internal/fuse/meta_dir.go
+++ b/internal/fuse/meta_dir.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package fuse

--- a/internal/fuse/other.go
+++ b/internal/fuse/other.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package fuse

--- a/internal/fuse/root.go
+++ b/internal/fuse/root.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package fuse

--- a/internal/fuse/snapshots_dir.go
+++ b/internal/fuse/snapshots_dir.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package fuse

--- a/internal/restic/node_solaris.go
+++ b/internal/restic/node_solaris.go
@@ -1,0 +1,27 @@
+package restic
+
+import "syscall"
+
+func (node Node) restoreSymlinkTimestamps(path string, utimes [2]syscall.Timespec) error {
+	return nil
+}
+
+func (s statUnix) atim() syscall.Timespec { return s.Atim }
+func (s statUnix) mtim() syscall.Timespec { return s.Mtim }
+func (s statUnix) ctim() syscall.Timespec { return s.Ctim }
+
+// Getxattr retrieves extended attribute data associated with path.
+func Getxattr(path, name string) ([]byte, error) {
+	return nil, nil
+}
+
+// Listxattr retrieves a list of names of extended attributes associated with the
+// given path in the file system.
+func Listxattr(path string) ([]string, error) {
+	return nil, nil
+}
+
+// Setxattr associates name and data together as an attribute of path.
+func Setxattr(path, name string, data []byte) error {
+	return nil
+}

--- a/internal/restic/node_xattr.go
+++ b/internal/restic/node_xattr.go
@@ -1,4 +1,5 @@
 // +build !openbsd
+// +build !solaris
 // +build !windows
 
 package restic

--- a/internal/restic/progress_unix.go
+++ b/internal/restic/progress_unix.go
@@ -1,4 +1,4 @@
-// +build !windows,!darwin,!freebsd,!netbsd,!openbsd,!dragonfly
+// +build !windows,!darwin,!freebsd,!netbsd,!openbsd,!dragonfly,!solaris
 
 package restic
 

